File: src/config/schema.ts
import { z } from "zod";

export const configSchema = z.object({
  gateway: z.object({
    port: z.number().default(19090),
    host: z.string().default("127.0.0.1"),
    authToken: z.string().optional(),
  }).default({}),

  agent: z.object({
    defaultProvider: z.enum(["openai", "anthropic"]).default("openai"),
    systemPrompt: z.string().default(
      "You are OpenAgent, a self-evolving personal AI assistant. " +
      "You can use tools to help the user, remember things across sessions, " +
      "create new skills, and even modify your own code. " +
      "Be concise, accurate, and proactive about learning from interactions."
    ),
    maxHistoryMessages: z.number().default(50),
    maxToolRounds: z.number().default(10),
  }).default({}),

  providers: z.object({
    openai: z.object({
      apiKey: z.string().default(""),
      baseUrl: z.string().default("https://api.openai.com/v1"),
      model: z.string().default("gpt-4o"),
      /** Extra query params appended to every request (e.g. { ak: "..." } for ByteDance GenAI) */
      queryParams: z.record(z.string()).default({}),
    }).default({}),
    anthropic: z.object({
      apiKey: z.string().default(""),
      model: z.string().default("claude-sonnet-4-20250514"),
    }).default({}),
  }).default({}),

  channels: z.object({
    webchat: z.object({
      enabled: z.boolean().default(true),
    }).default({}),
    feishu: z.object({
      enabled: z.boolean().default(false),
      appId: z.string().default(""),
      appSecret: z.string().default(""),
      encryptKey: z.string().default(""),
      verificationToken: z.string().default(""),
    }).default({}),
    telegram: z.object({
      enabled: z.boolean().default(false),
      botToken: z.string().default(""),
    }).default({}),
  }).default({}),

  evolution: z.object({
    memoryPath: z.string().default("./user-space/memory"),
    skillsPath: z.string().default("./user-space/skills"),
    selfModifyEnabled: z.boolean().default(true),
    reflectionEnabled: z.boolean().default(true),
  }).default({}),

  logging: z.object({
    level: z.enum(["debug", "info", "warn", "error"]).default("info"),
  }).default({}),

  storage: z.object({
    dbPath: z.string().default("./data/openagent.db"),
  }).default({}),
});

export type AppConfig = z.infer<typeof configSchema>;

File: src/config/schema.py
"""Configuration schema — Pydantic models matching the Zod schema in TS."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


def _to_camel(name: str) -> str:
    parts = name.split("_")
    return parts[0] + "".join(p.capitalize() for p in parts[1:])


class _CamelModel(BaseModel):
    """Base that reads/writes camelCase JSON while using snake_case in Python."""

    model_config = {"alias_generator": _to_camel, "populate_by_name": True}


class GatewayConfig(_CamelModel):
    port: int = 19090
    host: str = "127.0.0.1"
    auth_token: str | None = None


class OpenAIProviderConfig(_CamelModel):
    api_key: str = ""
    base_url: str = "https://api.openai.com/v1"
    model: str = "gpt-4o"
    query_params: dict[str, str] = Field(default_factory=dict)


class AnthropicProviderConfig(_CamelModel):
    api_key: str = ""
    model: str = "claude-opus-4-6"
    setup_token: str | None = None


class ProvidersConfig(_CamelModel):
    openai: OpenAIProviderConfig = Field(default_factory=OpenAIProviderConfig)
    anthropic: AnthropicProviderConfig = Field(default_factory=AnthropicProviderConfig)


class WebchatChannelConfig(_CamelModel):
    enabled: bool = True


class FeishuChannelConfig(_CamelModel):
    enabled: bool = False
    app_id: str = ""
    app_secret: str = ""
    encrypt_key: str = ""
    verification_token: str = ""


class TelegramChannelConfig(_CamelModel):
    enabled: bool = False
    bot_token: str = ""


class ChannelsConfig(_CamelModel):
    webchat: WebchatChannelConfig = Field(default_factory=WebchatChannelConfig)
    feishu: FeishuChannelConfig = Field(default_factory=FeishuChannelConfig)
    telegram: TelegramChannelConfig = Field(default_factory=TelegramChannelConfig)


class AgentConfig(_CamelModel):
    default_provider: Literal["openai", "anthropic", "claude-code"] = "openai"
    system_prompt: str = (
        "You are OpenAgent, a self-evolving personal AI assistant. "
        "You can use tools to help the user, remember things across sessions, "
        "create new skills, and even modify your own code. "
        "Be concise, accurate, and proactive about learning from interactions."
    )
    max_history_messages: int = 50
    max_tool_rounds: int = 10
    context_window: int = 128_000
    bootstrap_max_chars: int = 8_000
    bootstrap_total_max_chars: int = 40_000


class EvolutionConfig(_CamelModel):
    memory_path: str = "./user-space/memory"
    skills_path: str = "./user-space/skills"
    self_modify_enabled: bool = True
    reflection_enabled: bool = True


class LoggingConfig(_CamelModel):
    level: Literal["debug", "info", "warn", "error"] = "info"


class StorageConfig(_CamelModel):
    db_path: str = "./data/openagent.db"


class AppConfig(_CamelModel):
    gateway: GatewayConfig = Field(default_factory=GatewayConfig)
    agent: AgentConfig = Field(default_factory=AgentConfig)
    providers: ProvidersConfig = Field(default_factory=ProvidersConfig)
    channels: ChannelsConfig = Field(default_factory=ChannelsConfig)
    evolution: EvolutionConfig = Field(default_factory=EvolutionConfig)
    logging: LoggingConfig = Field(default_factory=LoggingConfig)
    storage: StorageConfig = Field(default_factory=StorageConfig)
